### ISSUE TITLE
Caffe2 and Caffe have different results with same input

### ISSUE THREAD

**Issue by tlhchutang** [facebookarchive/caffe2#595]
I have translated several caffemodels and prototxtes to corresponding pb files using the tool provided by Caffe2 successfully, however, when I was trying to test these pb files, I found that Caffe2 has quite different outputs (totally wrong to some extent) with those of Caffe using the same inputs, the input data of Caffe2 is aligned in the same order as that of Caffe. does anyone have any idea about this issue?

**Comment by lzx1413** [comment]
get the feature maps from responding layer and find out which layer is wrong

**Comment by tlhchutang** [comment]
found the reason.
In Caffe, there is no leakyReLU layer, we use ReLU with a negative slope parameter to implement leakyReLU;
However, in Caffe2's translator, it treats all ReLU  layers in Caffe prototxt as ReLU of Caffe2, including those with negative slopes, which are supposed to be treated as leakyReLU in Caffe2.
@Yangqing @bwasti

**Comment by tlhchutang** [comment]
by modifing caffe_translator.py, Caffe networks using LeakyReLU could be correctly translated into Caffe2. 

@TranslatorRegistry.Register("ReLU")
def TranslateRelu(layer, pretrained_blobs, is_test):
     param = layer.relu_param
     if param.negative_slope > 0:
         caffe_op = BaseTranslate(layer, "LeakyRelu")
         AddArgument(caffe_op, "alpha", param.negative_slope)
     else:
         caffe_op = BaseTranslate(layer, "Relu")
     return caffe_op, []

**Comment by nyyznyyz1991** [comment]
@lzx1413 @tlhchutang 
how to get the feature maps from responding layer? is there any python api?
i also face the same question, but i didn't use LeakyRelu, i use Batchnorm layer and add this layer's translate   script in caffe_translator.py, i can't figure out why the new translated caffe2 model doesn't work, the output feature is NAN or zero!!!

@TranslatorRegistry.Register("BatchNorm")
def TranslateBatchNorm(layer, pretrained_blobs, is_test):
    caffe_op = BaseTranslate(layer, "SpatialBN")
    output = caffe_op.output[0]

    param = layer.batch_norm_param
    AddArgument(caffe_op, "is_test", is_test)
    AddArgument(caffe_op, "epsilon", param.eps)
    AddArgument(caffe_op, "order", "NCHW")
    
    caffe_op.input.extend([output + "_scale", output + "_bias", output + "_mean", output + "_var"])
    if not is_test:
        caffe_op.output.extend([output + "_mean", output + "_var", output + "_saved_mean", output + "_saved_var"])
 
    n_channels = pretrained_blobs[0].shape[0] # get C
    mean = utils.NumpyArrayToCaffe2Tensor(pretrained_blobs[0], output + '_mean')
    var = utils.NumpyArrayToCaffe2Tensor(pretrained_blobs[1], output + '_var')
    pretrained_blobs[2] = np.tile(pretrained_blobs[2], (n_channels, )) # set C
    scale = utils.NumpyArrayToCaffe2Tensor(pretrained_blobs[2], output + '_scale')
    
    # Create a zero bias array the same size as the scale, we'll let the following
    # Scale (Mul + Add operators in Caffe2) layer handle any bias, just like Caffe
    bias = utils.NumpyArrayToCaffe2Tensor(np.zeros_like(pretrained_blobs[2]), output + '_bias')
    
    return caffe_op, [scale, bias, mean, var]

**Comment by tlhchutang** [comment]
@nyyznyyz1991 
you may try 
workspace.FeedBlob()
workspace.RunNet()
data= workspace.FetchBlob()

**Comment by spife129** [comment]
the fix has never been submitted to the github. Please merge the change! Also, negative_slope can be smaller than zero, the if statement should be changed.

**Comment by zhangw864680355** [comment]
@tlhchutang Hi,   I also successfully translator the caffe mode(tiny yolo) to init_net.pb and predict_net.pb file in caffe2.do you resolve the problem of  different results between caffe and caffe2 with same input ?
